src/hotspot/share/nmt/nmtTreap.hpp

```diff
@@ -0,0 +1,373 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_NMT_NMTTREAP_HPP
+#define SHARE_NMT_NMTTREAP_HPP
+
+#include "memory/allocation.hpp"
+#include "runtime/os.hpp"
+#include "utilities/debug.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include "utilities/growableArray.hpp"
+#include "utilities/macros.hpp"
+#include <stdint.h>
+
+// A Treap is a self-balanced binary tree where each node is equipped with a
+// priority. It adds the invariant that the priority of a parent P is strictly larger
+// larger than the priority of its children. When priorities are randomly
+// assigned the tree is balanced.
+// All operations are defined through merge and split, which are each other's inverse.
+// merge(left_treap, right_treap) => treap where left_treap <= right_treap
+// split(treap, key) => (left_treap, right_treap)  where left_treap <= right_treap
+// Recursion is used in these, but the depth of the call stack is the depth of
+// the tree which is O(log n) so we are safe from stack overflow.
+// TreapNode has LEQ nodes on the left, GT nodes on the right.
+//
+// COMPARATOR must have a static function `cmp(a,b)` which returns:
+//     - an int < 0 when a < b
+//     - an int == 0 when a == b
+//     - an int > 0 when a > b
+// ALLOCATOR must check for oom and exit, as Treap currently does not handle the allocation
+// failing.
+
+template<typename K, typename V, typename COMPARATOR, typename ALLOCATOR>
+class Treap {
+  friend class VMATreeTest;
+  friend class TreapTest;
+public:
+  class TreapNode {
+    friend Treap;
+    uint64_t _priority;
+    const K _key;
+    V _value;
+
+    TreapNode* _left;
+    TreapNode* _right;
+
+  public:
+    TreapNode(const K& k, const V& v, uint64_t p)
+      : _priority(p),
+        _key(k),
+        _value(v),
+        _left(nullptr),
+        _right(nullptr) {
+    }
+
+    const K& key() const { return _key; }
+    V& val() { return _value; }
+
+    TreapNode* left() const { return _left; }
+    TreapNode* right() const { return _right; }
+  };
+
+private:
+  ALLOCATOR _allocator;
+  TreapNode* _root;
+  uint64_t _prng_seed;
+  int _node_count;
+
+  uint64_t prng_next() {
+    // Taken directly off of JFRPrng
+    static const constexpr uint64_t PrngMult = 0x5DEECE66DLL;
+    static const constexpr uint64_t PrngAdd = 0xB;
+    static const constexpr uint64_t PrngModPower = 48;
+    static const constexpr uint64_t PrngModMask = (static_cast<uint64_t>(1) << PrngModPower) - 1;
+    _prng_seed = (PrngMult * _prng_seed + PrngAdd) & PrngModMask;
+    return _prng_seed;
+  }
```

Comment:
We are now adding a 3rd standalone identical implementation, there are 2 already in:

- ThreadHeapSampler::next_random()
- JfrPRNG::next_uniform()

Perhaps it's high time to consider moving this into `share/utilities` ?

Reply:
I agree, let's make a separate RFE for that. I'm keeping track of the instances where I've said that, so we shouldn't miss those opportunities :-).